src/tools/w3dviewer/scenelightdialog.cpp

```diff
@@ -0,0 +1,312 @@
+
+    if (doc != nullptr && doc->m_scene != nullptr) {
+        doc->m_light->Set_Diffuse(Vector3(m_diffuseRed / 100.0f, m_diffuseGreen / 100.0f, m_diffuseBlue / 100.0f));
+        doc->m_light->Set_Specular(Vector3(m_specularRed / 100.0f, m_specularGreen / 100.0f, m_specularBlue / 100.0f));
+        doc->m_light->Set_Intensity(m_intensity);
+        doc->m_light->Set_Far_Attenuation_Range(m_farAttenStart, m_farAttenEnd);
+        doc->m_light->Set_Flag(LightClass::FAR_ATTENUATION, m_attenuation == 1);
+        SetDistance(m_distance);
+    }
+
+    CDialog::OnCancel();
+}
+
+void CSceneLightDialog::OnHScroll(UINT nSBCode, UINT nPos, CScrollBar *pScrollBar)
+{
+    if (pScrollBar == GetDlgItem(IDC_INTENSITYSLIDER)) {
+        GetCurrentDocument()->m_light->Set_Intensity(m_intensitySlider.GetPos() / 100.0f);
+    } else {
+        if (SendDlgItemMessage(IDC_GREYSCALE, BM_GETCHECK)) {
+            HWND handle;
+
+            if (pScrollBar == GetDlgItem(IDC_REDSLIDER)) {
+                handle = m_redSlider.m_hWnd;
+            } else if (pScrollBar == GetDlgItem(IDC_GREENSLIDER)) {
+                handle = m_greenSlider.m_hWnd;
+            } else {
+                handle = m_blueSlider.m_hWnd;
+            }
+
+            int value = ::SendMessage(handle, TBM_GETPOS, 0, 0);
+            m_redSlider.SetPos(value);
+            m_greenSlider.SetPos(value);
+            m_blueSlider.SetPos(value);
+        }
+
+        CW3DViewDoc *doc = GetCurrentDocument();
+
+        if (doc != nullptr && doc->m_scene != nullptr) {
+            Vector3 color;
+            color.X = m_redSlider.GetPos() / 100.0f;
+            color.Y = m_greenSlider.GetPos() / 100.0f;
+            color.Z = m_blueSlider.GetPos() / 100.0f;
+
+            if ((m_flags & 2) != 0) {
+                doc->m_light->Set_Diffuse(color);
+            }
+
+            if ((m_flags & 4) != 0) {
+                doc->m_light->Set_Specular(color);
+            }
+        }
+    }
+
+    CWnd::OnHScroll(nSBCode, nPos, pScrollBar);
+}
+
+void CSceneLightDialog::OnGreyscale()
+{
+    if (SendDlgItemMessage(IDC_GREYSCALE, BM_GETCHECK) != 0) {
+        m_greenSlider.SetPos(m_redSlider.GetPos());
+        m_blueSlider.SetPos(m_redSlider.GetPos());
+    }
+
+    CW3DViewDoc *doc = GetCurrentDocument();
+
+    if (doc != nullptr && doc->m_scene != nullptr) {
+        Vector3 color;
+        color.X = m_redSlider.GetPos() / 100.0f;
+        color.Y = m_greenSlider.GetPos() / 100.0f;
+        color.Z = m_blueSlider.GetPos() / 100.0f;
+
+        if ((m_flags & 2) != 0) {
+            doc->m_light->Set_Diffuse(color);
+        }
+
+        if ((m_flags & 4) != 0) {
+            doc->m_light->Set_Specular(color);
+        }
+    }
+}
+
+void CSceneLightDialog::OnBoth()
+{
+    m_flags = 6;
+}
+
+void CSceneLightDialog::OnDiffuse()
+{
+    Vector3 color;
+    GetCurrentDocument()->m_light->Get_Diffuse(&color);
+
+    if (color.X == color.Y && color.X == color.Z) {
```

Comment:
This pattern is used in various conditions in this code.

I suggest to add inline helper function to improve readability
```c++
namespace
{
inline bool HasUniformChannels(color)
{
return color.X == color.Y && color.X == color.Z;
}
}
```